components/api-resource-mgt/org.wso2.carbon.identity.api.resource.collection.mgt/src/main/java/org/wso2/carbon/identity/api/resource/collection/mgt/APIResourceCollectionManagerImpl.java

```diff
@@ -0,0 +1,199 @@
+     *
+     * @param collectionId Collection id.
+     * @return API resource collection.
+     * @throws APIResourceCollectionMgtException If an error occurred while retrieving API Resource Collection.
+     */
+    @Override
+    public APIResourceCollection getAPIResourceCollectionById(String collectionId, String tenantDomain)
+            throws APIResourceCollectionMgtException {
+
+        APIResourceCollection apiResourceCollection = apiResourceCollectionMap.get(collectionId);
+        return populateAPIResourcesForCollection(apiResourceCollection, tenantDomain);
+    }
+
+    /**
+     * Populate API resources for the given API resource collection.
+     *
+     * @param collection   API resource collection.
+     * @param tenantDomain Tenant domain.
+     * @throws APIResourceCollectionMgtException If an error occurred while retrieving API resource metadata.
+     */
+    private APIResourceCollection populateAPIResourcesForCollection(APIResourceCollection collection,
+                                                                    String tenantDomain)
+            throws APIResourceCollectionMgtException {
+
+        if (collection == null) {
+            return null;
+        }
+        try {
+            APIResourceCollection clonedCollection = cloneAPIResourceCollection(collection);
+
+            // Combine read and write scopes for a single fetch.
+            Set<String> combinedScopes = new HashSet<>();
+            Optional.ofNullable(clonedCollection.getReadScopes()).ifPresent(combinedScopes::addAll);
+            Optional.ofNullable(clonedCollection.getWriteScopes()).ifPresent(combinedScopes::addAll);
+            List<APIResource> allAPIResources = APIResourceCollectionMgtServiceDataHolder.getInstance()
+                    .getAPIResourceManagementService().getScopeMetadata(new ArrayList<>(combinedScopes), tenantDomain);
+
+            List<APIResource> readAPIResources =
+                    filterAPIResources(allAPIResources, clonedCollection.getReadScopes());
+            List<APIResource> writeAPIResources =
+                    filterAPIResources(allAPIResources, clonedCollection.getWriteScopes());
+
+            Map<String, List<APIResource>> apiResourcesMap = new HashMap<>();
+            apiResourcesMap.put(APIResourceCollectionManagementConstants.READ, readAPIResources);
+            apiResourcesMap.put(APIResourceCollectionManagementConstants.WRITE, writeAPIResources);
+            clonedCollection.setApiResources(apiResourcesMap);
+            return clonedCollection;
+        } catch (APIResourceMgtException e) {
+            throw handleServerException(
+                    APIResourceCollectionManagementConstants.ErrorMessages
+                            .ERROR_CODE_ERROR_WHILE_RETRIEVING_SCOPE_METADATA,
+                    e);
+        }
+    }
+
+    private List<APIResource> filterAPIResources(List<APIResource> apiResources, List<String> scopes) {
+
+        if (scopes == null || scopes.isEmpty()) {
```

Comment:
Shall we use CollectionUtils.isEmpty(scopes)

Reply:
Fixed